lib/rspec/expectations/failure_aggregator.rb

```diff
@@ -0,0 +1,171 @@
+module RSpec
+  module Expectations
+    # @private
+    class FailureAggregator
+      attr_reader :block_label, :metadata
+
+      def aggregate
+        RSpec::Support.with_failure_notifier(self) do
+          begin
+            yield
+          rescue ExpectationNotMetError => e
+            # Normally, expectation failures will be notified via the `call` method, below,
+            # but since the failure notifier uses a thread local variable, failing expectations
+            # in another thread will still raise. We handle that here and categorize it as part
+            # of `failures` rather than letting it fall through and be categorized as part of
+            # `other_errors`.
+            failures << e
+          rescue Exception => e
+            # Normally we don't want to rescue `Exception` but here it's important.
+            # If we don't, it would cause expectation failures that happened before now to
+            # be silenced...but a failing precondition expressed as an expectation could point
+            # to the root cause of the exception we're rescuing. So it's important to include
+            # both in the raised error. `notify_aggregated_failures` will ensure that either
+            # just this exception (if there are no others) or a `MultipleExpectationsNotMetError`
+            # is raised as appropriate, so this exception will still be surfaced to the user.
```

Comment:
I don't think this covers why `Exception` and not `StandardError`? i.e. the same comment would apply if the code read `rescue => e`. That was the original thing that set my spidey sense off.